hyper: 3.0.2 -> 3.1.2 and fix for desktop files

FabianGD opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

I could not open hyper from the Applications menu on Gnome 34, as the executable in the `hyper.desktop` file was pointing to `/opt/Hyper/hyper`.

PS: This is my first-ever PR to nixpkgs, I hope everything is fine. 😄 

###### Things done

I fixed the hyper-generated desktop file (coming from the .deb package), as it was not pointing to the correct hyper executable. Updated hyper to the newest version 

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions (OpenSUSE Leap 15.3)
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

FabianGD committed:
(hyper): 3.0.2 -> 3.1.1 and fix for desktop files

r-rmcgibbo commented:
Result of `nixpkgs-review pr 132395` at 0554b39c run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>hyper</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/applications/terminal-emulators/hyper/default.nix:44:3:
```
   |
44 |   installPhase = ''
   |   ^
```
</li>
  </ul>
</details>

FabianGD committed:
hyper: fix deb hash
Fixes ownload of old cached deb package and addition of "new" dependencies.

FabianGD commented:
I just noticed that I changed the version but actually the deb package was pulled from the nixos.org instead because it seemingly only looked at the hash of the package. Now we actually pull version 3.1.1.

SuperSandro2000 commented on pkgs/applications/terminal-emulators/hyper/default.nix:
```diff
@@ -1,36 +1,58 @@
-    at-spi2-atk
+    at-spi2-atk at-spi2-core libxshmfence libdrm libxkbcommon mesa
   ];
+
+  name = "hyper";
+  desktopItem = makeDesktopItem {
+    type = "Application";
+    name = name;
+    desktopName = "Hyper";
+    genericName = "Hyper Terminal";
+    exec = name;
+    icon = "hyper";
+    categories = "System;TerminalEmulator;";
+  };
+
```
```suggestion
```

SuperSandro2000 commented on pkgs/applications/terminal-emulators/hyper/default.nix:
```diff
@@ -1,36 +1,58 @@
 stdenv.mkDerivation rec {
-  version = "3.0.2";
-  pname = "hyper";
+  pname = name;
```
```suggestion
  pname = "hyper";
```

SuperSandro2000 commented on pkgs/applications/terminal-emulators/hyper/default.nix:
```diff
@@ -1,36 +1,58 @@
+    copyDesktopItems ${desktopItem}/share/applications/* $out/share/applications/
   '';
+
+  desktopItems = [ desktopItem ];
```
```suggestion
  desktopItems = [
    (makeDesktopItem {
      type = "Application";
      name = "hyper";
      desktopName = "Hyper";
      genericName = "Hyper Terminal";
      exec = "hyper";
      icon = "hyper";
      categories = "System;TerminalEmulator;";
    })
  ];
```

FabianGD replied:
I inlined the desktopItem now, but did not store it in a singleton list now, as I copy the desktopItem manually in the installPhase (otherwise we will be left with the defective .desktop file from the deb package.)

SuperSandro2000 commented on pkgs/applications/terminal-emulators/hyper/default.nix:
```diff
@@ -1,36 +1,58 @@
+    url = "https://github.com/vercel/hyper/releases/download/v${version}/hyper_${version}_amd64.deb";
+    sha256 = "0j999z649k63rnr1lpbansaxnkrhrp73jdrdmslgnwbh7z1wsp9p";
   };
+  nativeBuildInputs = [ copyDesktopItems ];
+
   buildInputs = [ dpkg ];
```
```suggestion

  nativeBuildInputs = [ copyDesktopItems dpkg ];
```

FabianGD committed:
hyper: move desktopItem into mkDerivation

SuperSandro2000 commented on pkgs/applications/terminal-emulators/hyper/default.nix:
```diff
@@ -1,36 +1,57 @@
+    sha256 = "0j999z649k63rnr1lpbansaxnkrhrp73jdrdmslgnwbh7z1wsp9p";
+  };
+
+  buildInputs = [ copyDesktopItems dpkg ];
```
```suggestion
  nativeBuildInputs = [ copyDesktopItems dpkg ];
```

SuperSandro2000 commented on pkgs/applications/terminal-emulators/hyper/default.nix:
```diff
@@ -1,36 +1,57 @@
+
+  buildInputs = [ copyDesktopItems dpkg ];
+
+  desktopItems = makeDesktopItem {
+    type = "Application";
+    name = pname;
+    desktopName = "Hyper";
+    genericName = "Hyper Terminal";
+    exec = "hyper";
+    icon = "hyper";
+    categories = "System;TerminalEmulator;";
   };
```
```suggestion
  desktopItems = [
    (makeDesktopItem {
      type = "Application";
      name = pname;
      desktopName = "Hyper";
      genericName = "Hyper Terminal";
      exec = "hyper";
      icon = "hyper";
      categories = "System;TerminalEmulator;";
    })
  ];
```

SuperSandro2000 commented on pkgs/applications/terminal-emulators/hyper/default.nix:
```diff
@@ -1,36 +1,57 @@
     patchelf --set-interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" --set-rpath "${libPath}:$out/opt/Hyper:\$ORIGIN" "$out/opt/Hyper/hyper"
+
     mv usr/* "$out/"
+
+    copyDesktopItems ${desktopItems}/share/applications/* $out/share/applications/
   '';
```
```suggestion
  '';
```

SuperSandro2000 replied:
I didn't read your comment before.

> Did your suggestions work on your local installation?

I don't try these.

Can we maybe remove the desktop file from the deb package?

FabianGD replied:
I went a different route now by just replacing the `/opt/Hyper/hyper` hardcoded Exec path in the hyper.desktop file by `hyper`. What do you think about this solution?

FabianGD committed:
hyper: Simpler fix for the desktopItem issue
- Subsituting the erronous path for the executable name
- Cleanup of the function args

FabianGD commented:
3.1.2 brings some bugfixes and minor improvements, see [here](https://github.com/vercel/hyper/releases/tag/v3.1.2).
